File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        dp = [0] * (amount + 1)
        for i in range(1, len(dp)):
            curr = float('inf')
            for c in coins:
                if i >= c:
                    curr = min(curr, dp[i-c] + 1)
            dp[i] = curr

        return dp[amount] if dp[amount] != float('inf') else -1
                     

File: 0322-coin-change/0322-coin-change.java
class Solution {
    int[] dp = new int[10001];
    public int coinChange(int[] coins, int amount) {
    //     1,2,5   amount = 0.5
    //     / |  \
    //    1  2  5
    //  / | \
    // 1  2  5
        int res = backtrack(coins, amount);
        if (res == Integer.MAX_VALUE) return -1;
        return res;
    }

    private int backtrack(int[] coins, int amount) {
        if (amount == 0) return 0;
        if (amount < 0) return Integer.MAX_VALUE;
        if (dp[amount] > 0) return dp[amount];
        int curr = Integer.MAX_VALUE - 1;
        for (int i=0; i<coins.length; i++) {
            curr = Math.min(curr, backtrack(coins, amount - coins[i]));
        }
        return dp[amount] = curr + 1;
    }
}